02-microarray/pathway-analysis_microarray_03_gsva.Rmd

```diff
@@ -0,0 +1,428 @@
+In this case, it's also possible that a Entrez ID will map to multiple Ensembl IDs.
+For the purpose of performing GSEA later in this notebook, we keep only the first mapped IDs.
+For more about how to explore this, take a look at our [microarray gene ID conversion example](https://alexslemonade.github.io/refinebio-examples/02-microarray/gene-id-annotation_microarray_01_ensembl.html).
+
+Let's see a preview of `mapped_df`.
+
+```{r}
+head(mapped_df)
+```
+
+We will want to keep in mind that GSVA requires that data is in a matrix with the gene identifiers as rownames.
+In order to successfully turn our data frame into a matrix, we will need to ensure that we do not have any duplicate gene identifiers.
+
+Let's check to see if we have any Entrez IDs that mapped to multiple Ensembl IDs.
+
+```{r}
+any(duplicated(mapped_df$entrez_id))
+```
+
+Looks like we do have duplicated Entrez IDs.
+Let's find out which Entrez IDs have been duplicated.
+
+```{r}
+dup_entrez_ids <- mapped_df %>%
+  dplyr::filter(duplicated(entrez_id)) %>%
+  dplyr::pull(entrez_id)
+
+dup_entrez_ids
+```
+
+Now let's take a look at the rows associated with the duplicated Entrez IDs.
+
+```{r}
+mapped_df %>%
+  dplyr::filter(entrez_id %in% dup_entrez_ids)
+```
+
+We can see that the associated values vary for each row.
+
+As we mentioned earlier, we will want to remove duplicated gene identifiers in preparation for the GSVA steps later, so let's keep the Entrez IDs associated with the maximum expression value per sample.
+Note that using the maximum expression value to collapse duplicate gene identifiers is just one of the expression values that can be used here.
+See the [Gene Set Enrichment Analysis (GSEA) User guide](https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html) for more information on other expression values that can be used [@GSEA-user-guide]. 
+
+We are removing values for three gene identifiers here, so it is unlikely to have a considerable impact on our results.
+
+```{r}
+max_dup_df <- mapped_df %>%
+  # Filter to include only the rows associated with the duplicate gene identifiers
+  dplyr::filter(entrez_id %in% dup_entrez_ids) %>%
+  # Group by Entrez IDs
+  dplyr::group_by(entrez_id) %>%
+  # Get the maximum expression value using all values associated with the grouping variable
+  dplyr::summarize_all(max)
+
+max_dup_df
+```
+Looks like we were able to successfully get rid of the duplicate gene identifiers and keep the observations with the maximum expression value!
+
+Now let's combine the rest of the mapped data!
+
+```{r}
+filtered_mapped_df <- mapped_df %>%
+  dplyr::filter(
+    # First let's get the data associated with the gene identifiers that aren't duplicated
+    !entrez_id %in% dup_entrez_ids |
+      # Now let's get the data associated with the duplicated gene identifiers that we
+      # collapsed in `max_dup_df`
+      entrez_id %in% max_dup_df$entrez_id &
```

Comment:
As an example, I think this means you really are going to keep things based on what `max(c("ENSG00000107018", "ENSG00000107014"))` returns, which will just be based on character values of the id and not what you determined in `max_dup_df`.

Reply:
I think you will want to keep the first part of your filter (remove duplicated IDs) and then bind on `max_dup_df`. May want to do some kind of arrange after that so the dups aren't all at the end or beginning.

Reply:
I think to minimize some confusion, you can get rid of the `Ensembl` column earlier. Basically once your values are mixed up in `max_dup_df` the `Ensembl` column doesn't mean anything to us anymore anyway.

Reply:
Yes I see your point here. I could see how long it would take to run the steps on the whole mapped data frame rather than just on the data frame with the duplicated gene identifiers. That would be the ideal option here I believe (if not I'll try binding on `max_dup_df` again, it didn't work for me in development).

Reply:
>  Run the steps on the whole mapped data frame rather than just on the data frame with the duplicated gene identifiers. That would be the ideal option here I believe. 

Can you explain why you think this is ideal? (What if a user has less memory and a much bigger dataset -- I think this idea may be unnecessarily memory using)